### PR TITLE
UCS/CONFIG: Add infra for deprecated env vars

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

This PR implements the infrastructure to keep deprecated env variables and notify a user which tries to use them.
Also this PR restores UCX_MAX_SHORT and marks it as the deprecated one
Example:
```
$ UCX_UNKNWN=0 UCX_TCP_MAX_SHORT=1  UCX_MM_MAX_SHORT=1 ucx_info -eut
[1557931403.345563] [jazz01:76631:0]         parser.c:1517 UCX  WARN  unused env variable: UCX_UNKNWN (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
[1557931403.345575] [jazz01:76631:0]         parser.c:1523 UCX  WARN  deprecated env variables: (set UCX_WARN_UNUSED_ENV_VARS=n to suppress this warning)
[1557931403.345577] [jazz01:76631:0]         parser.c:1528 UCX  WARN  - UCX_TCP_MAX_SHORT: The configuration parameter replaced by: UCX_<IB transport>_TX_MIN_INLINE for IB, UCX_MM_FIFO_SIZE for MM
[1557931403.345579] [jazz01:76631:0]         parser.c:1528 UCX  WARN  - UCX_MM_MAX_SHORT: The configuration parameter replaced by: UCX_<IB transport>_TX_MIN_INLINE for IB, UCX_MM_FIFO_SIZE for MM
```

## Why ?

Will be used for #3524 

## How ?

Uses the same approach as for unused env vars (i.e. vars with "UCX_" prefix),:
1. adds deprecated env vars to the `used` list and then check whether this field is deprecated or not:
- if deprecated, it is added to the deprecated list and then a warning is printed out
2. if WARN_UNUSED env var set (by default, `yes`) it shows the warnings to the user

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10259/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10263/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10286/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> general question: is deprecated variable value ignored completely?

yes, they are ignored. Currently, we add a hint for a user to use a replacement for a deprecated variable (i.e. `UCX_MAX_SHORT` -> `UCX_MM_FIFO_SIZE` or `UCX_RC_MLX5_TX_MIN_INLINE`)

**Comment by yosefe** [comment]
@dmitrygladkov can we implement this simpler? maybe by a config parser which shows warning instead of storing the parsed value? or special value for "offset" e.g SIZE_MAX which will mean that it's deprecated?

**Comment by dmitrygladkov** [comment]
> can we implement this simpler? maybe by a config parser which shows warning instead of storing the parsed value? or special value for "offset" e.g SIZE_MAX which will mean that it's deprecated?

sure, good idea, it will add less memory overhead
will re-implement it

**Comment by dmitrygladkov** [comment]
> will re-implement it

@hoopoepg @brminich @yosefe 

force-pushed the new implementation

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10300/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10303/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10306/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10304/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10307/ for details (Mellanox internal link).

File: src/ucs/config/parser.c [hunk 1]
```diff
@@ -78,6 +88,29 @@ static int __find_string_in_list(const char *str, const char **list)
     return -1;
 }
 
+static int ucs_config_sscanf_deprecated(const char *buf, void *dest,
```
Comment (yosefe) [inline src/ucs/config/parser.c]: ucs_empty_function_do_assert()

File: src/ucs/config/parser.c [hunk 2]
```diff
@@ -795,12 +828,20 @@ void ucs_config_help_generic(char *buf, size_t max, const void *arg)
...
 }
 
-static inline int ucs_config_is_alias_field(const ucs_config_field_t *field)
+static inline unsigned
+ucs_config_is_deprecated_field(const ucs_config_field_t *field)
+{
+    return (field->offset == UCS_CONFIG_DEPRECATED_FIELD);
+}
+
+static inline unsigned
```
Comment (yosefe) [inline src/ucs/config/parser.c]: return int as previously

File: src/ucs/config/parser.c [hunk 3]
```diff
@@ -865,6 +906,14 @@ ucs_config_parser_set_default_values(void *opts, ucs_config_field_t *fields)
             continue;
         }
 
+        if (ucs_config_is_deprecated_field(field)) {
```
Comment (yosefe) [inline src/ucs/config/parser.c]: user has to set UCS_CONFIG_TYPE_DEPRECATED in the table

File: src/ucs/config/parser.c [hunk 4]
```diff
@@ -1222,15 +1276,16 @@ ucs_config_parser_print_field(FILE *stream, const void *opts, const char *env_pr
     ucs_assert(!ucs_list_is_empty(prefix_list));
     head = ucs_list_head(prefix_list, ucs_config_parser_prefix_t, list);
 
-    field->parser.write(value_buf, sizeof(value_buf) - 1, (char*)opts + field->offset,
```
Comment (yosefe) [inline src/ucs/config/parser.c]: if (deprecated) {
 snprintf(value_buf, " (deprecated)"
} else {
 snprintf(value_buf, "=")
 field->parser.write(value_buf +1, ...
}

and remove = from print below

File: test/gtest/common/main.cc [hunk 5]
```diff
@@ -74,16 +62,9 @@ int main(int argc, char **argv) {
     }
     UCS_TEST_MESSAGE << "Using random seed of " << ucs_gtest_random_seed;
     srand(ucs_gtest_random_seed);
+
+    ucs::init_config_for_valgrind(true);
```
Comment (yosefe) [inline test/gtest/common/main.cc]: move this to test_config only, save UCX_xx env vars , unset, test, and restore

Comment (dmitrygladkov) [reply test/gtest/common/main.cc]: done in the new PR that adds tests for unused env vars functionality - #3613

File: src/ucs/config/parser.c (hunk 1 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: done

File: src/ucs/config/parser.c (hunk 2 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: removed

File: src/ucs/config/parser.c (hunk 3 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: done

File: src/ucs/config/parser.c (hunk 4 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: good idea, done

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10326/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10328/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> Test FAILed.
> See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10328/ for details (Mellanox internal link).

it fails with:
```
...
12:52:46 OK ucs/profile/profile_defs.h
12:52:46 OK ucs/time/time_def.h
12:52:46 In file included from <stdin>:1:
12:52:46 ucs/config/parser.h:14:25: error: ucs/sys/sys.h: No such file or directory
```

@yosefe is it ok to rename `ucs_empty_function_do_assert()` to `ucs_config_parser_deprecated()`?
This is needed to not include `ucs/sys/sys.h` header file (it is not installed) in `ucs/config/parser.h` header file that is provided w/ UCX installation.

**Comment by yosefe** [comment]
need to put all these empty functions in new header+source files instead of sys.h/sys.c
e.g stubs.h + stubs.c

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10334/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10336/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> need to put all these empty functions in new header+source files instead of sys.h/sys.c
> e.g stubs.h + stubs.c

@yosefe done, added new files: `ucs/sys/stubs.h` and `ucs/sys/stubs.c`

File: src/ucs/config/parser.c [hunk 6]
```diff
@@ -1021,20 +1035,36 @@ static ucs_status_t ucs_config_apply_env_vars(void *opts, ucs_config_field_t *fi
...
+                if (ucs_config_is_deprecated_field(field)) {
+                    if ((status == UCS_OK) &&
+                        ucs_global_opts.warn_unused_env_vars) {
+                        ucs_warn("deprecated env variable %s used "
+                                 "(set %s%s=n to suppress this warning)",
+                                 buf, UCS_CONFIG_PREFIX,
+                                 UCS_GLOBAL_OPTS_WARN_UNUSED_CONFIG);
+                    }
+                } else {
+                    if ((status != UCS_OK) &&
```
Comment (yosefe) [inline src/ucs/config/parser.c]: why do we even need to check the status here? can just ignore it like before..

File: src/ucs/config/parser.c [hunk 7]
```diff
@@ -1222,9 +1254,16 @@ ucs_config_parser_print_field(FILE *stream, const void *opts, const char *env_pr
     ucs_assert(!ucs_list_is_empty(prefix_list));
     head = ucs_list_head(prefix_list, ucs_config_parser_prefix_t, list);
 
-    field->parser.write(value_buf, sizeof(value_buf) - 1, (char*)opts + field->offset,
-                        field->parser.arg);
-    field->parser.help(syntax_buf, sizeof(syntax_buf) - 1, field->parser.arg);
+    if (ucs_config_is_deprecated_field(field)) {
+        sprintf(value_buf, " (deprecated)");
```
Comment (yosefe) [inline src/ucs/config/parser.c]: always use s**n**printf

File: src/ucs/sys/compiler_def.h [hunk 8]
```diff
@@ -132,11 +132,23 @@
     })
 
 /**
- * @return Size of _member in _type. _type is a structure type.
+ * @param _type   Structure type.
```
Comment (yosefe) [inline src/ucs/sys/compiler_def.h]: unrelated change

File: src/tools/perf/api/libperf.h [hunk 9]
```diff
@@ -15,12 +15,13 @@ BEGIN_C_DECLS
...
 /** @file libperf.h */
 
-#include <sys/uio.h>
 #include <uct/api/uct.h>
 #include <ucp/api/ucp.h>
 #include <ucs/sys/math.h>
+#include <ucs/sys/stubs.h>
 #include <ucs/type/status.h>
 
+#include <sys/uio.h>
```
Comment (yosefe) [inline src/tools/perf/api/libperf.h]: unrelated change

File: src/tools/perf/lib/libperf.c [hunk 10]
```diff
@@ -10,6 +10,7 @@
 #include <ucs/debug/log.h>
 #include <ucs/arch/bitops.h>
 #include <ucs/sys/module.h>
+
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c]: unrelated change

File: src/ucs/sys/stubs.c [hunk 11]
```diff
@@ -0,0 +1,68 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include <ucs/sys/stubs.h>
+#include <ucs/debug/assert.h>
+
```
Comment (yosefe) [inline src/ucs/sys/stubs.c]: pls add another space line

File: src/ucs/config/parser.h [hunk 12]
```diff
@@ -182,6 +184,7 @@ void ucs_config_help_table(char *buf, size_t max, const void *arg);
 void ucs_config_release_nop(void *ptr, const void *arg);
 void ucs_config_help_generic(char *buf, size_t max, const void *arg);
 
+#define UCS_CONFIG_DEPRECATED_FIELD SIZE_MAX
```
Comment (yosefe) [inline src/ucs/config/parser.h]: UCS_CONFIG_DEPRECATED_FIELD -> UCS_CONFIG_DEPRECATED_FIELD_OFFSET

File: src/tools/perf/api/libperf.h (hunk 9 above)
Comment (dmitrygladkov) [reply src/tools/perf/api/libperf.h]: removed

File: src/tools/perf/lib/libperf.c (hunk 10 above)
Comment (dmitrygladkov) [reply src/tools/perf/lib/libperf.c]: removed

File: src/ucs/config/parser.c [hunk 13]
```diff
@@ -1021,20 +1035,36 @@ static ucs_status_t ucs_config_apply_env_vars(void *opts, ucs_config_field_t *fi
...
-                ucs_config_parser_mark_env_var_used(buf);
-                status = ucs_config_parser_parse_field(field, env_value, var);
-                if (status != UCS_OK) {
-                    /* If set to ignore errors, restore the default value */
-                    ucs_status_t tmp_status =
-                        ucs_config_parser_parse_field(field, field->dfl_value,
-                                                      var);
-                    if (ignore_errors) {
-                        status = tmp_status;
+                status = ucs_config_parser_mark_env_var_used(buf);
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: done

File: src/ucs/config/parser.c (hunk 6 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: I think that Coverity will be not happy with that, because we will set the new value `status` variable w/o touching/using it before
```
status = ucs_config_parser_mark_env_var_used(buf);
if ((status != UCS_OK) &&
    (status != UCS_ERR_ALREADY_EXISTS)) {
    ucs_warn(...);
    continue;
}

... // the same code
```

File: src/ucs/config/parser.c [hunk 14]
```diff
@@ -1021,20 +1035,36 @@ static ucs_status_t ucs_config_apply_env_vars(void *opts, ucs_config_field_t *fi
...
+
+                if (ucs_config_is_deprecated_field(field)) {
+                    if ((status == UCS_OK) &&
+                        ucs_global_opts.warn_unused_env_vars) {
+                        ucs_warn("deprecated env variable %s used "
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: done

File: src/ucs/config/parser.c [hunk 15]
```diff
@@ -1021,20 +1035,36 @@ static ucs_status_t ucs_config_apply_env_vars(void *opts, ucs_config_field_t *fi
...
-                    ucs_status_t tmp_status =
-                        ucs_config_parser_parse_field(field, field->dfl_value,
-                                                      var);
-                    if (ignore_errors) {
-                        status = tmp_status;
+                status = ucs_config_parser_mark_env_var_used(buf);
+
+                if (ucs_config_is_deprecated_field(field)) {
+                    if ((status == UCS_OK) &&
+                        ucs_global_opts.warn_unused_env_vars) {
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: done

File: src/ucs/config/parser.c (hunk 7 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: done

File: src/ucs/config/parser.h (hunk 12 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.h]: done

File: src/ucs/sys/compiler_def.h (hunk 8 above)
Comment (dmitrygladkov) [reply src/ucs/sys/compiler_def.h]: removed

File: src/ucs/config/parser.c (hunk 6 above)
Comment (dmitrygladkov) [reply src/ucs/config/parser.c]: discussed offline with @yosefe:
will refactor `ucs_config_parser_mark_env_var_used` to return the value from the argument and check inside `if` and don't check inside `else`

File: src/ucs/sys/stubs.c (hunk 11 above)
Comment (dmitrygladkov) [reply src/ucs/sys/stubs.c]: done

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10362/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe @brminich this is ready for the next round of review
